File: mobile/src/main/java/net/maiatoday/levelbest/ui/MainActivity.java
package net.maiatoday.levelbest.ui;

import android.content.Intent;
import android.content.SharedPreferences;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.Snackbar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.view.Menu;
import android.view.MenuItem;

import com.google.firebase.analytics.FirebaseAnalytics;

import net.maiatoday.levelbest.LevelBestApplication;
import net.maiatoday.levelbest.R;
import net.maiatoday.levelbest.databinding.ActivityMainBinding;
import net.maiatoday.levelbest.helpers.PreferenceHelper;

import javax.inject.Inject;

public class MainActivity extends AppCompatActivity {
    @Inject
    SharedPreferences prefs;

    @Inject
    FirebaseAnalytics analytics;

    private boolean firstTime;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((LevelBestApplication)getApplication()).getComponent().inject(this);
        firstTime = prefs.getBoolean(PreferenceHelper.KEY_FIRST_TIME, true);
        PreferenceHelper.write(prefs, PreferenceHelper.KEY_FIRST_TIME, false);
        analytics.logEvent(FirebaseAnalytics.Event.APP_OPEN, null);
        ActivityMainBinding binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        Toolbar toolbar = binding.toolbar;
        setSupportActionBar(toolbar);

        FloatingActionButton fab = binding.fab;
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Bundle bundle = new Bundle();
                bundle.putString(FirebaseAnalytics.Param.ITEM_ID, "id");
                bundle.putString(FirebaseAnalytics.Param.ITEM_NAME, "test");
                bundle.putString(FirebaseAnalytics.Param.CONTENT_TYPE, "image");
                analytics.logEvent(FirebaseAnalytics.Event.SELECT_CONTENT, bundle);
                if (firstTime) {
                    Snackbar.make(view, "First Time - Replace with your own action "+utils.DemoUtils.addTest(2,4), Snackbar.LENGTH_LONG)
                            .setAction("Action", null).show();
                } else {
                    Snackbar.make(view, "Replace with your own action "+utils.DemoUtils.addTest(4,4), Snackbar.LENGTH_LONG)
                            .setAction("Action", null).show();

                }
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            startActivity(new Intent(this, SettingsActivity.class));
            return true;
        }

        return super.onOptionsItemSelected(item);
    }
}

File: mobile/src/main/kotlin/net/maiatoday/levelbest/view/MainActivity.kt
package net.maiatoday.levelbest.view

import android.content.Intent
import android.content.SharedPreferences
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.design.widget.BottomNavigationView
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.view.View
import com.google.firebase.analytics.FirebaseAnalytics
import io.realm.Realm
import io.realm.Sort
import net.maiatoday.levelbest.LevelBestApplication
import net.maiatoday.levelbest.R
import net.maiatoday.levelbest.databinding.ActivityMainBinding
import net.maiatoday.levelbest.helpers.PreferenceHelper
import net.maiatoday.levelbest.model.Entry
import net.maiatoday.levelbest.view.adapters.EntryRecyclerAdapter
import javax.inject.Inject

class MainActivity : AppCompatActivity(), EntryRecyclerAdapter.OnEntryClick {

    @Inject
    lateinit var prefs: SharedPreferences

    @Inject
    lateinit var analytics: FirebaseAnalytics

    private var firstTime: Boolean = false

    companion object {
        private val REQUEST_NEW_ENTRY = 100
        private val REQUEST_OLD_ENTRY = 101
    }

    private lateinit var realm: Realm

    private val mOnNavigationItemSelectedListener = BottomNavigationView.OnNavigationItemSelectedListener { item ->
        when (item.itemId) {
            R.id.navigation_moods -> {
                startActivity(MoodsActivity.makeIntent(this))
                return@OnNavigationItemSelectedListener true
            }
            R.id.navigation_tags -> {
                startActivity(TagsActivity.makeIntent(this))
                return@OnNavigationItemSelectedListener true
            }
            R.id.navigation_settings -> {
                startActivity(Intent(this, SettingsActivity::class.java))
                return@OnNavigationItemSelectedListener true
            }
        }
        false
    }

    private lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (application as LevelBestApplication).component.inject(this)
        realm = Realm.getDefaultInstance()
        firstTime = prefs.getBoolean(PreferenceHelper.KEY_FIRST_TIME, true)
        PreferenceHelper.write(prefs, PreferenceHelper.KEY_FIRST_TIME, false)

        analytics.logEvent(FirebaseAnalytics.Event.APP_OPEN, null)
        binding = DataBindingUtil.setContentView<ActivityMainBinding>(this, R.layout.activity_main)
        val toolbar = binding.toolbar
        setSupportActionBar(toolbar)

        val navigation = binding.navigation
        navigation.setOnNavigationItemSelectedListener(mOnNavigationItemSelectedListener)

        setUpRecyclerView()

        val fab = binding.fab
        fab.setOnClickListener { view ->
            val bundle = Bundle()
            bundle.putString(FirebaseAnalytics.Param.ITEM_ID, "id")
            bundle.putString(FirebaseAnalytics.Param.ITEM_NAME, "test")
            bundle.putString(FirebaseAnalytics.Param.CONTENT_TYPE, "image")
            analytics.logEvent(FirebaseAnalytics.Event.SELECT_CONTENT, bundle)
            startActivityForResult(EntryActivity.makeIntent(this, ""), REQUEST_NEW_ENTRY)
        }
    }

    private fun setUpRecyclerView() {
        binding.entriesContent.list.layoutManager = LinearLayoutManager(this)
        binding.entriesContent.list.adapter = EntryRecyclerAdapter(this,
                this,
                realm.where(Entry::class.java).findAllSortedAsync(Entry.TIMESTAMP, Sort.ASCENDING))
        binding.entriesContent.list.setHasFixedSize(true)
        binding.entriesContent.list.addItemDecoration(DividerItemDecoration(this, DividerItemDecoration.VERTICAL))
    }

    override fun entryClick(view: View, data: Entry) {
        startActivityForResult(EntryActivity.makeIntent(this, data.id), REQUEST_OLD_ENTRY)
    }

    override fun onDestroy() {
        realm.close()
        super.onDestroy()
    }
}
